File: app/src/main/java/io/mainframe/hacs/PageFragments/StatusFragment.java
package io.mainframe.hacs.PageFragments;

import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.EnumSet;

import io.mainframe.hacs.R;
import io.mainframe.hacs.common.YesNoDialog;
import io.mainframe.hacs.components.DoorButtons;
import io.mainframe.hacs.main.BackDoorStatus;
import io.mainframe.hacs.main.NetworkStatus;
import io.mainframe.hacs.main.Status;
import io.mainframe.hacs.mqtt.MqttConnector;
import io.mainframe.hacs.mqtt.MqttStatusListener;
import io.mainframe.hacs.ssh.DoorCommand;
import io.mainframe.hacs.ssh.PkCredentials;

import static io.mainframe.hacs.common.Constants.SPACE_DOOR;

public class StatusFragment extends BasePageFragment implements NetworkStatus.NetworkStatusListener, MqttStatusListener {

    private DoorButtons doorButtons;

    public StatusFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_status, container, false);

        doorButtons = (DoorButtons) view.findViewById(R.id.status_doorButtons);
        doorButtons.setOnButtonClickListener(new DoorButtons.OnButtonClickListener() {
            @Override
            public void onClick(final DoorButtons.DoorButton doorButton, View view) {
                if (doorButton.getStatus() == Status.CLOSE) {
                    final BackDoorStatus backDoorStatus = getInteraction().getMqttConnector().getLastValue(Topic.BACK_DOOR_BOLT, BackDoorStatus.class);
                    if (backDoorStatus == BackDoorStatus.OPEN) {
                        YesNoDialog.show(getContext(), "Back-Door prüfen.", "Die Back-Door ist noch offen. Wirklich abschließen?", "bd", new YesNoDialog.ResultListener() {
                            @Override
                            public void dialogClosed(String tag, boolean resultOk) {
                                if (resultOk) {
                                    getInteraction().sendSshCommand(SPACE_DOOR, DoorCommand.getSwitchDoorStateCmd(doorButton.getStatus()));
                                }
                            }
                        });

                        return;
                    }
                }

                getInteraction().sendSshCommand(SPACE_DOOR, DoorCommand.getSwitchDoorStateCmd(doorButton.getStatus()));
            }
        });

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();

        PkCredentials credentials = new PkCredentials(PreferenceManager.getDefaultSharedPreferences(getActivity()));
        // if the ssh key password is not set
        boolean readOnlyMode = !credentials.isPasswordSet();

        if (!readOnlyMode) {
            final NetworkStatus networkStatus = getInteraction().getNetworkStatus();
            networkStatus.addListener(this);
            doorButtons.setEnabled(networkStatus.isInMainframeWifi() && !networkStatus.hasMachiningBssid());
        } else {
            doorButtons.setEnabled(false);
        }

        final MqttConnector mqtt = getInteraction().getMqttConnector();
        mqtt.addListener(this, EnumSet.of(Topic.STATUS, Topic.BACK_DOOR_BOLT));

        setStatusText(mqtt.getLastValue(Topic.STATUS, Status.class));
        setLedImage(mqtt.getLastValue(Topic.BACK_DOOR_BOLT, BackDoorStatus.class));
    }

    @Override
    public void onPause() {
        super.onPause();

        getInteraction().getMqttConnector().removeAllListener(this);
        getInteraction().getNetworkStatus().removeListener(this);
    }

    @Override
    public int getTitleRes() {
        return R.string.nav_status;
    }

    private void setStatusText(Status status) {
        TextView text = getView().findViewById(R.id.status_status);
        text.setText(status == null ? getString(R.string.unknown) : status.getUiValue());
    }

    private void setLedImage(BackDoorStatus status) {
        final ImageView imageView = getView().findViewById(R.id.back_door_status);

        switch (status) {
            case OPEN:
                imageView.setImageResource(R.drawable.ic_led_red_black);
                break;
            case CLOSED:
                imageView.setImageResource(R.drawable.ic_led_blue_black);
                break;
            default:
                imageView.setImageResource(R.drawable.ic_button_black);
        }
    }

    /* callback */

    @Override
    public void onNetworkChange(boolean hasNetwork, boolean hasMobile, boolean hasWifi, boolean isInMainframeWifi, boolean hasMachiningBssid) {
        doorButtons.setEnabled(isInMainframeWifi);
    }

    @Override
    public void onNewMsg(Topic topic, Object msg) {
        if (topic == Topic.STATUS) {
            setStatusText((Status) msg);
        } else if (topic == Topic.BACK_DOOR_BOLT) {
            setLedImage((BackDoorStatus) msg);
        }
    }

    @Override
    public void onMqttConnected() {
        // not needed
    }

    @Override
    public void onMqttConnectionLost() {
        setStatusText(null);
        setLedImage(BackDoorStatus.UNKNOWN);
    }
}

File: app/src/main/kotlin/io/mainframe/hacs/PageFragments/StatusFragment.kt
package io.mainframe.hacs.PageFragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import io.mainframe.hacs.R
import io.mainframe.hacs.common.Constants.SPACE_DOOR
import io.mainframe.hacs.common.YesNoDialog
import io.mainframe.hacs.components.DoorButtons
import io.mainframe.hacs.main.BackDoorStatus
import io.mainframe.hacs.main.NetworkStatus
import io.mainframe.hacs.main.Status
import io.mainframe.hacs.ssh.DoorCommand
import io.mainframe.hacs.ssh.PkCredentials
import io.mainframe.hacs.status.StatusEvent
import io.mainframe.hacs.status.Subscription
import io.mainframe.hacs.trash_notifications.TrashCalendar
import org.pmw.tinylog.Logger

class StatusFragment : BasePageFragment(), NetworkStatus.NetworkStatusListener {

    private var subscription: Subscription? = null
    private lateinit var doorButtons: DoorButtons

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_status, container, false)

        doorButtons = view.findViewById(R.id.status_doorButtons)
        doorButtons.setOnButtonClickListener { doorButton, _ ->
            if (doorButton.status == Status.CLOSE) {
                // special action when the space is going to be closed
                withBackDoorCheck {
                    withTrashCheck {
                        interaction.sendSshCommand(
                            SPACE_DOOR,
                            DoorCommand.getSwitchDoorStateCmd(doorButton.status)
                        )
                    }
                }

            } else {
                interaction.sendSshCommand(
                    SPACE_DOOR,
                    DoorCommand.getSwitchDoorStateCmd(doorButton.status)
                )
            }
        }

        return view
    }

    private fun withTrashCheck(next: () -> Unit) {
        val trashSummaryForTomorrow = TrashCalendar(context).trashSummaryForTomorrow
        if (trashSummaryForTomorrow == null) {
            next()
            return
        }

        YesNoDialog.show(
            context, "Müll prüfen",
            String.format(
                "Morgen ist Müllabfuhr! Ist schon %s an die Straße gestellt?",
                trashSummaryForTomorrow
            ),
            "trash"
        ) { _, resultOk ->
            if (resultOk) {
                next()
            }
        }
    }

    private fun withBackDoorCheck(next: () -> Unit) {
        if (getBackdoorStatus() != BackDoorStatus.OPEN) {
            next()
            return
        }

        YesNoDialog.show(
            context,
            "Back-Door prüfen.",
            "Die Back-Door ist noch offen. Wirklich abschließen?",
            "bd"
        ) { _, resultOk ->
            if (resultOk) {
                next()
            }
        }
    }

    override fun onResume() {
        super.onResume()

        // if the ssh key password is not set
        val readOnlyMode = !PkCredentials.isPasswordSet(context!!)

        if (!readOnlyMode) {
            val networkStatus = interaction.networkStatus
            networkStatus.addListener(this)

            if (networkStatus.isRequireMainframeWifi) {
                doorButtons.isEnabled =
                    networkStatus.isInMainframeWifi && !networkStatus.hasMachiningBssid()
            } else {
                doorButtons.isEnabled = true
            }
        } else {
            doorButtons.isEnabled = false
        }

        val statusService = interaction.statusService
        subscription = statusService.subscribe { event: StatusEvent, value: String ->
            if (event == StatusEvent.SPACE_STATUS) {
                activity?.runOnUiThread {
                    setStatusText(Status.byEventStatusValue(value))
                }
            }
        }

        setStatusText(statusService.getLastStatusValue(StatusEvent.SPACE_STATUS))
        setLedImage(getBackdoorStatus())
    }

    override fun onPause() {
        super.onPause()

        subscription?.unsubscribe()
        interaction.networkStatus.removeListener(this)
    }

    override fun getTitleRes(): Int {
        return R.string.nav_status
    }

    private fun setStatusText(status: Status?) {
        val view = view ?: return
        val text = view.findViewById<TextView>(R.id.status_status)
        text.text = if (status == null) getString(R.string.unknown) else status.uiValue
    }

    private fun getBackdoorStatus(): BackDoorStatus {
        return try {
            interaction.statusService.getLastValue(StatusEvent.BACKDOOR)
                ?.let { BackDoorStatus.byMqttValue(it) }
                ?: BackDoorStatus.UNKNOWN
        } catch (e: IllegalStateException) {
            Logger.warn("Unexpected Backdoor status: ${e.message}")
            BackDoorStatus.UNKNOWN
        }
    }

    private fun setLedImage(status: BackDoorStatus?) {
        val view = view ?: return

        val imageView = view.findViewById<ImageView>(R.id.back_door_status)
        when (status) {
            BackDoorStatus.OPEN -> imageView.setImageResource(R.drawable.ic_led_red_black)
            BackDoorStatus.CLOSED -> imageView.setImageResource(R.drawable.ic_led_blue_black)
            else -> imageView.setImageResource(R.drawable.ic_button_black)
        }
    }

    /* callback */

    override fun onNetworkChange(
        hasNetwork: Boolean, hasMobile: Boolean, hasWifi: Boolean,
        isInMainframeWifi: Boolean, hasMachiningBssid: Boolean, requireMainframeWifi: Boolean
    ) {
        doorButtons.isEnabled = !requireMainframeWifi || isInMainframeWifi
    }
}
